HW2/P4/driver.py

```diff
@@ -0,0 +1,169 @@
+import sys
+import os.path
+sys.path.append(os.path.join('..', 'util'))
+
+import set_compiler
+set_compiler.install()
+
+import pyximport
+pyximport.install()
+
+import numpy as np
+import pylab
+
+import filtering
+from timer import Timer
+import threading
+
+
+# ORIGINAL CODE
+# def py_median_3x3(image, iterations=10, num_threads=1):
+#     ''' repeatedly filter with a 3x3 median '''
+#     tmpA = image.copy()
+#     tmpB = np.empty_like(tmpA)
+    
+
+#     for i in range(iterations):
+#         filtering.median_3x3(tmpA, tmpB, 0, 1)
+#         # swap direction of filtering
+#         tmpA, tmpB = tmpB, tmpA
+
+#     return tmpA
+
+
+#Function that ensure that threads do not start an interation of filtering before the data for that 
+#iteration is ready
+def parallel_function(tmpA, tmpB, t, num_threads, events_list, iterations):
+    for i in range(iterations):
+        #We clear the event t
+        events_list[t].clear()
+        
+        #We filter the image (only the rows t, t+4, t+4+4, etc)
+        filtering.median_3x3(tmpA, tmpB, t, num_threads)
+        
+        #One the filtering is done we put a flag
+        events_list[t].set()
+        
+        #We swap direction of filtering
+        tmpA, tmpB = tmpB, tmpA
+        
+        for e in events_list:
+            #We make sure that we are not waiting for ourselves
+            if e!=t:
+                #This blocks the events to go forward until all the events of that iteration are done
+                e.wait()
+
+#Extra credit. Function that allows threads to begin an interation of filtering asynchronously
```

Comment:
You are one of the few that I have seen to get the asynchronous portion to work, good work.